File: backend/src/main/java/com/back/domain/bookmarks/repository/BookmarkRepositoryCustomImpl.java
package com.back.domain.bookmarks.repository;

import com.back.domain.bookmarks.constant.ReadState;
import com.back.domain.bookmarks.dto.ReadStateCount;
import com.back.domain.bookmarks.entity.Bookmark;
import com.back.domain.member.member.entity.Member;
import com.querydsl.core.BooleanBuilder;
import com.querydsl.core.Tuple;
import com.querydsl.core.types.Order;
import com.querydsl.core.types.OrderSpecifier;
import com.querydsl.core.types.dsl.PathBuilder;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.back.domain.book.author.entity.QAuthor.author;
import static com.back.domain.book.book.entity.QBook.book;
import static com.back.domain.book.wrote.entity.QWrote.wrote;
import static com.back.domain.bookmarks.entity.QBookmark.bookmark;

@RequiredArgsConstructor
public class BookmarkRepositoryCustomImpl implements BookmarkRepositoryCustom {
    private final JPAQueryFactory queryFactory;

    @Override
    public Page<Bookmark> search(Member member, String category, String readState, String keyword, Pageable pageable) {
        BooleanBuilder builder = conditions(member, category, readState, keyword);

        List<Bookmark> bookmarks = queryFactory
                .select(bookmark).distinct()
                .from(bookmark)
                .join(bookmark.book, book)
                .leftJoin(book.authors, wrote)
                .leftJoin(wrote.author, author)
                .where(builder)
                .offset(pageable.getOffset())
                .limit(pageable.getPageSize())
                .orderBy(getOrderSpecifiers(pageable.getSort()))
                .fetch();

        Long total = queryFactory
                .select(bookmark.countDistinct())
                .from(bookmark)
                .join(bookmark.book, book)
                .leftJoin(book.authors, wrote)
                .leftJoin(wrote.author, author)
                .where(builder)
                .fetchOne();
        return new PageImpl<>(bookmarks, pageable, total != null ? total : 0L);
    }

    private OrderSpecifier<?>[] getOrderSpecifiers(Sort sort){
        if(sort.isEmpty()){
            return new OrderSpecifier[]{bookmark.createDate.desc()};
        }
        return sort.stream().map(order -> {
            Order direction = order.isAscending() ? Order.ASC : Order.DESC;
            String property = order.getProperty();
            PathBuilder<Bookmark> pathBuilder = new PathBuilder<>(Bookmark.class, "bookmark");
            return new OrderSpecifier(direction, pathBuilder.get(property));
        })
                .toArray(OrderSpecifier[]::new);
    }

    @Override
    public ReadStateCount countReadState(Member member, String category, String readState, String keyword) {
        BooleanBuilder builder = conditions(member, category, readState, keyword);

        List<Tuple> counts = queryFactory
                .select(bookmark.readState, bookmark.countDistinct())
                .from(bookmark)
                .join(bookmark.book, book)
                .leftJoin(book.authors, wrote)
                .leftJoin(wrote.author, author)
                .where(builder)
                .groupBy(bookmark.readState)
                .fetch();

        Map<ReadState, Long> countMap = counts.stream().collect(Collectors.toMap(
                tuple -> tuple.get(0, ReadState.class),
                tuple -> tuple.get(1, Long.class)
        ));

        ReadStateCount readStateCount = new ReadStateCount(
                countMap.getOrDefault(ReadState.READ, 0L),
                countMap.getOrDefault(ReadState.READING, 0L),
                countMap.getOrDefault(ReadState.WISH, 0L)
        );
        return readStateCount;
    }

    private BooleanBuilder conditions(Member member, String category, String readState, String keyword) {
        BooleanBuilder builder = new BooleanBuilder();
        builder.and(bookmark.member.eq(member));

        if(category != null && !category.isBlank()){
            builder.and(bookmark.book.category.name.eq(category));
        }

        if(readState != null && !readState.isBlank()){
            builder.and(bookmark.readState.eq(ReadState.valueOf(readState.toUpperCase())));
        }

        if(keyword != null && !keyword.isBlank()){
            builder.and(
                    book.title.containsIgnoreCase(keyword)
                            .or(author.name.containsIgnoreCase(keyword))
            );
        }

        return builder;
    }
}

File: backend/src/main/java/com/back/domain/bookmarks/repository/BookmarkRepositoryCustomImpl.kt
package com.back.domain.bookmarks.repository

import com.back.domain.book.author.entity.QAuthor
import com.back.domain.book.book.entity.QBook
import com.back.domain.book.wrote.entity.QWrote
import com.back.domain.bookmarks.constant.ReadState
import com.back.domain.bookmarks.dto.ReadStateCount
import com.back.domain.bookmarks.entity.Bookmark
import com.back.domain.bookmarks.entity.QBookmark
import com.back.domain.member.member.entity.Member
import com.back.global.jpa.querydsl.QuerydslUtil
import com.querydsl.core.BooleanBuilder
import com.querydsl.core.Tuple
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.support.PageableExecutionUtils
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors


class BookmarkRepositoryCustomImpl(
    private val queryFactory: JPAQueryFactory
) : BookmarkRepositoryCustom {
    override fun search(
        member: Member,
        category: String?,
        state: String?,
        keyword: String?,
        pageable: Pageable
    ): Page<Bookmark> {
        val builder = conditions(member, category, state, keyword)
        val specifiers = QuerydslUtil.toOrderSpecifiers(pageable.getSort(), QBookmark.bookmark)
        val bookmarks = queryFactory
            .select(QBookmark.bookmark).distinct()
            .from(QBookmark.bookmark)
            .join(QBookmark.bookmark.book, QBook.book)
            .leftJoin(QBook.book.authors, QWrote.wrote)
            .leftJoin(QWrote.wrote.author, QAuthor.author)
            .where(builder)
            .offset(pageable.getOffset())
            .limit(pageable.getPageSize().toLong())
            .orderBy(*specifiers.toTypedArray())
            .fetch()

        val total = queryFactory
            .select(QBookmark.bookmark.countDistinct())
            .from(QBookmark.bookmark)
            .join(QBookmark.bookmark.book, QBook.book)
            .leftJoin(QBook.book.authors, QWrote.wrote)
            .leftJoin(QWrote.wrote.author, QAuthor.author)
            .where(builder)

        return PageableExecutionUtils.getPage(bookmarks, pageable) { total.fetchOne()?: 0L }
    }

    override fun countReadState(
        member: Member,
        category: String?,
        state: String?,
        keyword: String?
    ): ReadStateCount {
        val builder = conditions(member, category, state, keyword)
        val counts = queryFactory!!
            .select(QBookmark.bookmark.readState, QBookmark.bookmark.countDistinct())
            .from(QBookmark.bookmark)
            .join(QBookmark.bookmark.book, QBook.book)
            .leftJoin(QBook.book.authors, QWrote.wrote)
            .leftJoin(QWrote.wrote.author, QAuthor.author)
            .where(builder)
            .groupBy(QBookmark.bookmark.readState)
            .fetch()
        val countMap = counts.stream().collect(
            Collectors.toMap(
                Function { tuple: Tuple? -> tuple!!.get(0, ReadState::class.java) },
                Function { tuple: Tuple? -> tuple!!.get<Long?>(1, Long::class.java) }
            ))

        val readStateCount = ReadStateCount(
            countMap.getOrDefault(ReadState.READ, 0L)?:0L,
            countMap.getOrDefault(ReadState.READING, 0L)?:0L,
            countMap.getOrDefault(ReadState.WISH, 0L)?:0L
        )
        return readStateCount
    }



    private fun conditions(member: Member, category: String?, readState: String?, keyword: String?): BooleanBuilder {
        val builder = BooleanBuilder()
        builder.and(QBookmark.bookmark.member.eq(member))

        if (category != null && !category.isBlank()) {
            builder.and(QBookmark.bookmark.book.category.name.eq(category))
        }

        if (readState != null && !readState.isBlank()) {
            builder.and(QBookmark.bookmark.readState.eq(ReadState.valueOf(readState.uppercase(Locale.getDefault()))))
        }

        if (keyword != null && !keyword.isBlank()) {
            builder.and(
                QBook.book.title.containsIgnoreCase(keyword)
                    .or(QAuthor.author.name.containsIgnoreCase(keyword))
            )
        }

        return builder
    }
}
